### PR TITLE
[FSDP] Allow MixedPrecision to skip inputs

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90621
* __->__ #90620
* #90579
* #90523

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
LGTM, but curious what's the use case to skip inputs?

**Comment by mrshenli** [comment]
Thanks @rohan-varma, one model I am working on has one forward argument, which is sensitive to precision. So have to keep it in fp32, instead of converting it to bfloat16 and convert back.

File: torch/distributed/fsdp/api.py [hunk 1]
```diff
@@ -117,6 +117,9 @@ class MixedPrecision:
             gradients back to the full parameter precision after the backward
             pass. This may be set to ``False`` to save memory if using custom
             optimizers that can perform the optimizer step in ``reduce_dtype``.
+            (Default: ``False``)
+        apply_to_forward_inputs (bool): Apply ``param_dtype`` to forward
+            arguments and keyward arguments. (Default: ``True``)
```
Comment (awgu) [inline torch/distributed/fsdp/api.py:122]:
```suggestion
            arguments and keyword arguments. (Default: ``True``)
```

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 2]
```diff
@@ -832,6 +832,52 @@ def test_float16_on_one_submodule(self):
...
+
+        fsdp(x).sum().backward()
+
+        self.assertEqual(forward_inputs[model].dtype, torch.float32)
+        self.assertEqual(forward_inputs[c1].dtype, torch.float32)
+        self.assertEqual(forward_inputs[c2].dtype, torch.float32)
+
+    @skip_if_lt_x_gpu(2)
+    def test_float16_on_one_submodule_skip_inputs_error(self):
+        forward_inputs: Dict[str, nn.Module] = {}
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:860]:
```suggestion
        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
```

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 3]
```diff
@@ -832,6 +832,52 @@ def test_float16_on_one_submodule(self):
         self.assertEqual(forward_inputs[c1].dtype, torch.float32)
         self.assertEqual(forward_inputs[c2].dtype, torch.float16)
 
+    @skip_if_lt_x_gpu(2)
+    def test_float16_on_one_submodule_skip_inputs(self):
+        forward_inputs: Dict[str, nn.Module] = {}
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:837]:
```suggestion
        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
```

File: torch/distributed/fsdp/api.py [hunk 4]
```diff
@@ -117,6 +117,9 @@ class MixedPrecision:
             gradients back to the full parameter precision after the backward
             pass. This may be set to ``False`` to save memory if using custom
             optimizers that can perform the optimizer step in ``reduce_dtype``.
+            (Default: ``False``)
+        apply_to_forward_inputs (bool): Apply ``param_dtype`` to forward
```
Comment (awgu) [inline torch/distributed/fsdp/api.py:121]: nit: I am wondering if maybe we should call this `cast_forward_inputs`?

Though, I totally understand what you mean by `apply_to_forward_inputs` -- i.e. "are we applying mixed precision to the forward inputs?".

This is just a suggestion to consider. I think `cast_forward_inputs` might be more explicit.

---

```suggestion
        cast_forward_inputs (bool): Cast floating point tensors in the forward arguments and keyword arguments to ``param_dtype``.
```

File: torch/testing/_internal/common_distributed.py [hunk 5]
```diff
@@ -957,21 +957,30 @@ def world_size(self) -> int:
 
 
 class SaveForwardInputsModule(nn.Module):
-    def __init__(self, forward_inputs: Dict[nn.Module, torch.Tensor]) -> None:
+    def __init__(
+        self,
+        forward_inputs: Dict[nn.Module, torch.Tensor],
+        convert_inputs: bool = False,
```
Comment (awgu) [inline torch/testing/_internal/common_distributed.py:963]: nit: Let us make this non-defaulted since we always pass it explicitly in the unit tests?

**Review by awgu — APPROVED** [review]
Overall, this looks good to me. Sorry for leaving all of the comments in separate reviews instead of just one.

**Comment by awgu** [comment]
My bad for suggesting to un-default `cast_forward_inputs`. I did not see that it already existed in past tests.

I think you need to add a value for `cast_forward_inputs` in this line:
https://github.com/pytorch/pytorch/blob/a5d34ad0c51273ec4e008a777252c18f240ca8fb/test/distributed/fsdp/test_fsdp_mixed_precision.py#L821

**Pull request closed without merging**